bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,926 @@
+     *
+     * @throws IOException
+     *
+     */
+    public synchronized void stopCommunication() {
+        listenerStopped = true;
+
+        Socket socket = qSocket;
+
+        if (socket != null) {
+            try {
+                socket.close();
+            } catch (IOException ignore) {
+                // ignore IO Error when trying to close the socket if the intention is to close it anyway
+            }
+        }
+
+        qSocket = null;
+
+        CTDConnected = false;
+        logger.debug("Communication stopped from thread {}", Thread.currentThread().getId());
+    }
+
+    /**
+     * Close and restart communication with Qbus Server.
+     */
+    public synchronized void restartCommunication() {
+        stopCommunication();
+
+        logger.debug("Qbus: restart communication from thread {}", Thread.currentThread().getId());
+
+        startCommunication();
+    }
+
+    /**
+     * Method to check if communication with Qbus Server is active
+     *
+     * @return True if active
+     */
+    public boolean communicationActive() {
+
+        return qSocket != null;
+    }
+
+    /**
+     * Method to check if communication with Qbus Client is active
+     *
+     * @return True if active
+     */
+
+    public boolean clientConnected() {
+        return CTDConnected;
+    }
+
+    /**
+     * Runnable that handles incomming communication from Qbus server.
+     * <p>
+     * The thread listens to the TCP socket opened at instantiation of the {@link QbusCommunication} class
+     * and interprets all incomming json messages. It triggers state updates for active channels linked to the
+     * Qbus outputs. It is started after initialization of the communication.
+     *
+     */
+    private Runnable qEvents = () -> {
+        String qMessage;
+
+        logger.info("Listening for events on thread {}", Thread.currentThread().getId());
+        listenerStopped = false;
+        qEventsRunning = true;
+
+        BufferedReader reader = qIn;
+
+        try {
+            if (reader == null) {
+                throw new IOException();
+            }
+            while (!listenerStopped & ((qMessage = reader.readLine()) != null)) {
+                if (qMessage != null) {
+                    readMessage(qMessage);
+                }
+            }
+        } catch (IOException e) {
+            if (!listenerStopped) {
+                qEventsRunning = false;
+                logger.warn("Qbus: IO error in listener on thread {}", Thread.currentThread().getId());
+
+                QbusBridgeHandler handler = bridgeCallBack;
+
+                if (handler != null) {
+                    CTDConnected = false;
+                    handler.bridgeOffline("No communication with Qbus server");
+                }
+
+                return;
+            }
+        }
+
+        qEventsRunning = false;
+
+        logger.warn("Event listener thread stopped on thread {}", Thread.currentThread().getId());
```

Comment:
Stopping a thread is a normal operation when OH is shut-down. You could make this debug or trace.